### PR TITLE
[READY] playsound performance

### LINKED ISSUE #40131: Emitters can be heard on separate Z Levels
[Round ID]: 93504

Was ghosting at Charlie Station and from RnD I could hear emitters. An admin confirmed that I was at the same X and Y coordinates as the emitters on the main station Z level.

### PR DESCRIPTION
**Pull request by vuonojenmustaturska**
Closes #40131

### CODE REVIEW

**Comment by vuonojenmustaturska** [comment]
image
FREE PERFORMANCE

**Comment by vuonojenmustaturska** [comment]
tl;dr: we have lists of playermobs on the z-level, which generally have fewer player mobs to iterate through. If we use them, we can get rid of the z-level checks too. On the off-chance that something plays a sound in nullspace, we can skip pretty much everything. The client-check can go too, there are no playermobs without a client.

**Comment by DaxDupont** [comment]
I can't hear sounds anymore as a ghost, like door sounds and what not.

File: code/game/sound.dm [hunk 1]
```diff
@@ -5,25 +5,20 @@
...
-			continue
-		var/distance = get_dist(M, turf_source)
-
-		if(distance <= maxdistance)
-			var/turf/T = get_turf(M)
-
-			if(T && T.z == turf_source.z)
+	if (turf_source)
+		//allocate a channel if necessary now so its the same for everyone
+		channel = channel || open_sound_channel()
```
Comment (yoyobatty) [inline code/game/sound.dm]: shouldn't this be channel |= open_sound_channel() ?

Comment (SpaceManiac) [reply code/game/sound.dm]: No, `|` and `||` are different. It could be `||=` if we had that but we don't.

**Comment by optimumtact** [comment]
@DaxDupont feature am I rite

**Comment by JohnGinnane** [comment]
@DaxDupont looks like 
https://github.com/tgstation/tgstation/blob/b109515f64c2886b9f25e4e308cf225da1ba003f/code/modules/mob/living/living.dm#L1034-L1050
would need to be run on `mob/` rather than `mob/living`

File: code/game/sound.dm [hunk 2]
```diff
@@ -5,25 +5,20 @@
...
-		var/mob/M = P
-		if(!M || !M.client)
-			continue
-		var/distance = get_dist(M, turf_source)
-
-		if(distance <= maxdistance)
-			var/turf/T = get_turf(M)
-
-			if(T && T.z == turf_source.z)
+	if (turf_source)
```
Comment (ninjanomnom) [inline code/game/sound.dm]:
```
if(!turf_source)
  return
```
save your local coder and reduce whitespace today

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by vuonojenmustaturska** [comment]
image
lost some performance to no longer skipping ghosts

File: code/modules/admin/verbs/adminjump.dm [hunk 3]
```diff
@@ -66,9 +66,8 @@
 
 	if(src.mob)
 		var/mob/A = src.mob
-		A.x = tx
-		A.y = ty
-		A.z = tz
```
Comment (Cyberboss) [inline code/modules/admin/verbs/adminjump.dm]: Nigga hwat?

**Comment by vuonojenmustaturska** [comment]
ugh, preeetty sure i fixed what would've caused #40131 in the latest commit but there is no round id, let's see if another one report pops up

File: code/game/sound.dm [hunk 4]
```diff
@@ -5,26 +5,27 @@
...
-		var/distance = get_dist(M, turf_source)
-
-		if(distance <= maxdistance)
-			var/turf/T = get_turf(M)
-
-			if(T && T.z == turf_source.z)
-				M.playsound_local(turf_source, soundin, vol, vary, frequency, falloff, channel, pressure_affected, S)
+		if(get_dist(M, turf_source) <= maxdistance)
+			M.playsound_local(turf_source, soundin, vol, vary, frequency, falloff, channel, pressure_affected, S)
+	for(var/P in SSmobs.dead_players_by_zlevel[z])
```
Comment (Fox-McCloud) [inline code/game/sound.dm]: @vuonojenmustaturska 

I realize this system is no longer in place, but wouldn't `clients_by_zlevel` contains any dead players on the same z-level?

What's the purpose of looping through both?

**Pull request merged**